src/git-unix/ogit-http-fetch/main.ml

```diff
@@ -0,0 +1,169 @@
+    | Logs.Info -> Logs_fmt.info_style
+  in
+  let level = Logs.level_to_string (Some level) in
+  Fmt.pf ppf "[%a][%a]"
+    (Fmt.styled level_style Fmt.string)
+    level (Fmt.option Fmt.string)
+    (Option.map (pad 10) header)
+
+let reporter ppf =
+  let report src level ~over k msgf =
+    let k _ =
+      over ();
+      k ()
+    in
+    let with_src_and_stamp h _ k fmt =
+      let dt = Mtime.Span.to_us (Mtime_clock.elapsed ()) in
+      Fmt.kpf k ppf
+        ("%s %a %a: @[" ^^ fmt ^^ "@]@.")
+        (pad 10 (Fmt.str "%+04.0fus" dt))
+        pp_header (level, h)
+        Fmt.(styled `Magenta string)
+        (pad 10 @@ Logs.Src.name src)
+    in
+    msgf @@ fun ?header ?tags fmt -> with_src_and_stamp header tags k fmt
+  in
+  { Logs.report }
+
+let setup_logs style_renderer level ppf =
+  Fmt_tty.setup_std_outputs ?style_renderer ();
+  Logs.set_level level;
+  Logs.set_reporter (reporter ppf);
+  let quiet = match style_renderer with Some _ -> true | None -> false in
+  quiet, ppf
+
+type error = [ `Store of Store.error | `Sync of Sync.error ]
+
+let store_err err = `Store err
+let sync_err err = `Sync err
+
+let pp_error ppf = function
+  | `Store err -> Fmt.pf ppf "(`Store %a)" Store.pp_error err
+  | `Sync err -> Fmt.pf ppf "(`Sync %a)" Sync.pp_error err
+
+let main (references : (Git.Reference.t * Git.Reference.t) list)
+    (directory : string option) (repository : Uri.t) =
+  let ( >>?= ) = Lwt_result.bind in
+  let ( >>!= ) v f = Lwt_result.map_err f v in
+  let root : Fpath.t =
+    Fpath.v @@ match directory with Some p -> p | None -> Sys.getcwd ()
+  in
+  let resolvers : Conduit.resolvers =
+    Conduit_lwt.empty
+    |> Conduit_lwt.add Conduit_lwt.TCP.protocol
+         (Conduit_lwt.TCP.resolve ~port:80)
+  in
+  let remote_references = List.map fst references in
+  Lwt.return @@ Smart_git.endpoint_of_string (Uri.to_string repository)
+  >>!= store_err
+  >>?= fun endpoint ->
+  Store.v root >>!= store_err >>?= fun store ->
+  Sync.fetch ~resolvers endpoint store (`Some remote_references) >>!= sync_err
+
+open Cmdliner
+
+module Flag = struct
+  let output_value =
+    let parse str =
+      match str with
+      | "stdout" -> Ok Fmt.stdout
+      | "stderr" -> Ok Fmt.stderr
+      | s -> Error (`Msg (Fmt.strf "%s is not an output." s))
+    in
+    let print ppf v =
+      Fmt.pf ppf "%s" (if v == Fmt.stdout then "stdout" else "stderr")
+    in
+    Arg.conv ~docv:"<output>" (parse, print)
+
+  let output =
+    let doc = "Output of the progress status" in
+    Arg.(
+      value
+      & opt output_value Fmt.stdout
+      & info [ "output" ] ~doc ~docv:"<output>")
+
+  let progress =
+    let doc =
+      "Progress status is reported on the standard error stream by default \
+       when it is attached to a terminal, unless -q is specified. This flag \
+       forces progress status even if the standard error stream is not \
+       directed to a terminal."
+    in
+    Arg.(value & flag & info [ "progress" ] ~doc)
```

Comment:
Unfortunately, from what I did about `Smart_git`, it's not possible (but should be easy to extend the API) to print what the server does. Indeed, when we do a `fetch`, the server is able to _write_ something to the terminal such as:
```
remote: Enumerating objects: 193, done.
remote: Total 193 (delta 0), reused 0 (delta 0), pack-reused 193
```

Such outputs is available here:
https://github.com/mirage/ocaml-git/blob/436e84defab9b41c6024d5462584e27fa3fb5f33/src/not-so-smart/protocol.ml#L549-L558

But I just decided to ignore them:

https://github.com/mirage/ocaml-git/blob/436e84defab9b41c6024d5462584e27fa3fb5f33/src/not-so-smart/fetch.ml#L118-L119

 As a good introspection of `Smart_git` it can be good to extend the API to be able to print such message.